pkg/lib/bundle/validate.go

```diff
@@ -170,6 +192,64 @@ func (i imageValidator) ValidateBundleFormat(directory string) error {
 	return nil
 }
 
+func validateProperties(manifestsDir string, fileBasedProperties []PropertiesTypeVal) ([]error, []string) {
+	var validationErrors []error
+	var validationWarnings []string
+
+	// Read the manifests directory for the CSV
+	manifestFiles, err := ioutil.ReadDir(manifestsDir)
+	if err != nil {
+		validationErrors = append(validationErrors, fmt.Errorf("Failed to read manifests directory while validating properties: %w", err))
+	}
+
+	// Find the CSV
+	csvFound := false
+	csvFile := &v1.ClusterServiceVersion{}
+	for _, f := range manifestFiles {
+		err = registry.DecodeFile(filepath.Join(manifestsDir, f.Name()), &csvFile)
+		if err == nil && csvFile.Kind == CSVKind {
+			csvFound = true
+			break
+		}
+	}
+
+	// If CSV is found and it has the olm.properties attribute defined
+	if csvFound {
+		if annotationPropertiesString, defined := csvFile.Annotations["olm.properties"]; defined {
+			// Structure the olm.properties into a comparable format
+			var annotationBasedProperties []PropertiesTypeVal
+			err := json.Unmarshal([]byte(annotationPropertiesString), &annotationBasedProperties)
+			if err != nil {
+				validationErrors = append(validationErrors, fmt.Errorf("Failed to read CSV file while validating properties: %w", err))
+			}
+
+			// Determine which props have more values
+			moreProperties := annotationBasedProperties
+			lessProperties := fileBasedProperties
+			if len(moreProperties) < len(lessProperties) {
+				moreProperties = fileBasedProperties
+				lessProperties = annotationBasedProperties
+			}
+
+			// For each property in both slices, determine if a property is duplicated, then warn if not done already
+			alreadyWarned := map[string]bool{}
+			for _, mp := range moreProperties {
+				for _, lp := range lessProperties {
+					if mp.Type == lp.Type && !alreadyWarned[mp.Type] {
+						alreadyWarned[mp.Type] = true
+						validationWarnings = append(
+							validationWarnings,
+							fmt.Sprintf("Found property %s defined in both CSV and properties.yaml", mp.Type),
```

Comment:
are we also checking for defining a property multiple times within the same source/defining the property multiple times with the same value?

Reply:
This is a good point. Thinking on it, I would say that the main case to worry about (that we don't already cover in this PR) is if the same source has a key/value pair that is identical. It would be good to have a separate warning in that case which shouldn't be too hard to act on. We don't care so much if two different sources have the exact same key/value pair as they're already warned on due to the key. Would you agree?

Reply:
Alright, this logic should now be implemented. It also has unit test cases to accompany it.